api/src/controllers/elastic-search.ts

```diff
@@ -0,0 +1,130 @@
+} from "../types";
+import {
+  CountResponse,
+  ErrorCause,
+} from "@elastic/elasticsearch/lib/api/types";
+import { Config, getConfig } from "../../config";
+import { articlesContentTypes, graphQueryArticles } from "../helpers/articles";
+import { transformArticle } from "../transformers/article";
+import { HttpError } from "./error";
+import { PaginationQueryParameters } from "./pagination";
+import { RequestHandler } from "express";
+
+const indexName = getConfig().contentsIndex;
+
+type QueryParams = {
+  query?: string;
+} & PaginationQueryParameters;
+
+type Handler = RequestHandler<
+  never,
+  { count: CountResponse },
+  never,
+  QueryParams
+>;
+
+// TODO remove?
+const addIndex = async (elasticClient: Client, indexName: string) => {
+  const exists = await elasticClient.indices.exists({ index: indexName });
+
+  if (!exists) {
+    await elasticClient.indices.create({ index: indexName });
+    console.log(indexName, "was created");
+  } else {
+    console.log("Index", indexName, "already exists");
+  }
+};
+
+const useBulkHelper = async (elasticClient: Client, docs: Article[]) => {
+  const operations = docs.flatMap((doc) => {
+    return [
+      { index: { _index: indexName, _id: doc.id } },
+      {
+        display: doc as Article,
+        query: { id: doc.id },
+      },
+    ];
+  });
+
+  const bulkResponse = await elasticClient.bulk({ refresh: true, operations });
+
+  if (bulkResponse.errors) {
+    const erroredDocuments: {
+      status: number;
+      error: ErrorCause;
+      operation: any;
+      document: any;
+    }[] = [];
+    // The items array has the same order of the dataset we just indexed.
+    // The presence of the `error` key indicates that the operation
+    // that we did for the document has failed.
+    bulkResponse.items.forEach((action, i) => {
+      if (action["index"]?.error) {
+        erroredDocuments.push({
+          // If the status is 429 it means that you can retry the document,
+          // otherwise it's very likely a mapping error, and you should
+          // fix the document before to try it again.
+          status: action["index"]?.status,
+          error: action["index"]?.error,
+          operation: operations[i * 2],
+          document: operations[i * 2 + 1],
+        });
+      }
+    });
+    console.log("erroredDocuments", erroredDocuments);
+  }
+
+  const count = await elasticClient.count({ index: indexName });
+  return count;
+};
+
+const elasticSearchController = (clients: Clients, config: Config): Handler => {
+  return async (req, res) => {
+    try {
+      const searchResponse = await clients.prismic.getAllByType<
+        ArticlePrismicDocument & {
+          contentType: ContentType | ContentType[];
+        }
+      >("articles", {
+        // We'll need to consider "webcomics" as well when we do this properly
+        graphQuery: graphQueryArticles,
+        predicates: [
+          prismic.predicate.any("document.type", articlesContentTypes),
```

Comment:
I can't remember if I asked this before - why do we have a predicate for `document.type` _and_ a `getAllByType`? Isn't the latter doing the type filtering?